openshift-201/rh201-pod-auto-scale.md

```diff
@@ -0,0 +1,283 @@
+
+### Advanced Options
+
+There are a few more advanced options with the HPAs like scaleup and scaledown policies. Check the online documentation for these details.
+
+* https://docs.openshift.com/container-platform/4.8/nodes/pods/nodes-pods-autoscaling.html
+* https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/
+
+
+## Vertical Pod Autoscaler
+
+VPA reduces resources for pods that are requesting more resources than they are using and increases resources for pods that are not requesting enough.
+
+The VPA reviews historic and current CPU and memory resources for containers in pods and can update the resource limits and requests based on the usage values it learns. The VPA will update all of the pods associated with a workload object, such as a Deployment, DeploymentConfig, StatefulSet, Job, DaemonSet, ReplicaSet, or ReplicationController, in a project.
+
+VPA automatically deletes any pods that are out of alignment with its recommendations one at a time, so that your applications can continue to serve requests with no downtime. By default, workload objects must specify a minimum of two replicas in order for the VPA to automatically delete their pods. Workload objects that specify fewer replicas than this minimum are not deleted. If you manually delete these pods, when the workload object redeploys the pods, the VPA does update the new pods with its recommendations.
+
+For developers, you can use the VPA to help ensure your pods stay up during periods of high demand by scheduling pods onto nodes that have appropriate resources for each pod. Administrators use the VPA to better utilize cluster resources, such as preventing pods from reserving more CPU resources than needed. The VPA monitors the resources that workloads are actually using and adjusts the resource requirements so capacity is available to other workloads. 
+
+The VPA CR must be in the same project as the pods you want to monitor. There are 4 updateModes available for the VPA:
+
+* Auto, the VPA assigns resource requests on pod creation and updates the existing pods by terminating them when the requested resources differ significantly from the new recommendation.
+* Recreate, the VPA assigns resource requests on pod creation and updates the existing pods by terminating them when the requested resources differ significantly from the new recommendation. This mode should be used rarely, only if you need to ensure that the pods are restarted whenever the resource request changes. Otherwise prefer the "Auto" mode which may take advantage of restart free updates once they are available.
+* Initial, automatically applies VPA recommendations only at pod creation.
+* Off, only provides recommended resource limits and requests, allowing you to manually apply the recommendations. The off mode does not update pods.
+
+
+Lets first confirm that the VPA operator is installed in our cluster, we should see 2 results.
+
+```
+oc api-resources | grep vpa
```

Comment:
```suggestion
oc api-resources |grep vpa
```

Reply:
That space doesn't really matter. I'm just going to leave it.

Reply:
When I copied the command with the space included, I couldn't run it and it took me a while to figure out what was going on. Maybe this is specific to my command line interface?

Reply:
I think so. A space should work fine with most bash shells